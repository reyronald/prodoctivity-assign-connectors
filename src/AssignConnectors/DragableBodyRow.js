// @flow
import * as React from "react";
import { DragSource, DropTarget } from "react-dnd";
import { css } from "react-emotion";

function dragDirection(
  dragIndex,
  hoverIndex,
  initialClientOffset,
  clientOffset,
  sourceClientOffset
): "downward" | "upward" | void {
  const hoverMiddleY = (initialClientOffset.y - sourceClientOffset.y) / 2;
  const hoverClientY = clientOffset.y - sourceClientOffset.y;
  if (dragIndex < hoverIndex && hoverClientY > hoverMiddleY) {
    return "downward";
  }
  if (dragIndex > hoverIndex && hoverClientY < hoverMiddleY) {
    return "upward";
  }
}

const downward = css`
  td {
    border-bottom: 2px dashed #1890ff !important;
  }
`;

const upward = css`
  td {
    border-top: 2px dashed #1890ff;
  }
`;

class BodyRow extends React.Component {
  render() {
    const {
      isOver,
      connectDragSource,
      connectDropTarget,
      moveRow,
      dragRow,
      clientOffset,
      sourceClientOffset,
      initialClientOffset,
      ...restProps
    } = this.props;
    const style = { ...restProps.style, cursor: "move" };

    let className = restProps.className;
    if (isOver && initialClientOffset) {
      const direction = dragDirection(
        dragRow.index,
        restProps.index,
        initialClientOffset,
        clientOffset,
        sourceClientOffset
      );
      if (direction === "downward") {
        className += " " + downward;
      }
      if (direction === "upward") {
        className += " " + upward;
      }
    }
    return connectDragSource(
      connectDropTarget(
        <tr {...restProps} className={className} style={style} />
      )
    );
  }
}

const rowSource = {
  beginDrag(props) {
    return {
      index: props.index
    };
  }
};

const rowTarget = {
  drop(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Time to actually perform the action
    props.moveRow(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex;
  }
};

const DragableBodyRow = DropTarget("row", rowTarget, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOver: monitor.isOver(),
  sourceClientOffset: monitor.getSourceClientOffset()
}))(
  DragSource("row", rowSource, (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    dragRow: monitor.getItem(),
    clientOffset: monitor.getClientOffset(),
    initialClientOffset: monitor.getInitialClientOffset()
  }))(BodyRow)
);

export default DragableBodyRow;
