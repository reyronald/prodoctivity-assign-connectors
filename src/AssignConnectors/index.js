// @flow
export { default } from "./AssignConnectors";
