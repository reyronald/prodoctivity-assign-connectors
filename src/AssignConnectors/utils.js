// @flow

import type { WebServiceConnector, FormConnector } from "../_lib/types";
import type { ConnectorToAssign } from "./AssignConnectors";

export function mergeConnectorsAndFormConnectors(
  connectors: Array<WebServiceConnector>,
  formConnectors: Array<FormConnector> = []
): Array<ConnectorToAssign> {
  const _formConnectors = formConnectors.slice(0);
  const connectorsToAssign = connectors.map(connector => {
    const formConnector = pluck(
      _formConnectors,
      fc => fc.name === connector.name
    );

    if (!formConnector) {
      return {
        key: connector.name,
        name: connector.name,
        output: [],
        shouldDisableOnData: true,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      };
    }

    return {
      selected: true,
      key: formConnector.name,
      name: formConnector.name,
      output: formConnector.output.filter(o =>
        connector.outputMapping.some(om => om.dataElement.name === o.name)
      ),
      shouldDisableOnData:
        formConnector.shouldDisableOnData == null
          ? true
          : formConnector.shouldDisableOnData,
      shouldDisableOnNoData:
        formConnector.shouldDisableOnNoData == null
          ? false
          : formConnector.shouldDisableOnNoData,
      noDataMessage: formConnector.noDataMessage || ""
    };
  });

  return connectorsToAssign;
}

function pluck<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => boolean
): ?T {
  const index = array.findIndex(predicate);
  if (index > -1) {
    return array.splice(index, 1)[0];
  }
}
