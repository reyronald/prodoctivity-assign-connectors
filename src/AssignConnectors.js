// @flow
import React from "react";
import { Table, Checkbox, Icon, Divider, Radio, Input } from "antd";
import { filter, match } from "fuzzaldrin-plus";
import { splitMatches } from "./utils";

const rowSelection = {
  onChange: (selectedRowKeys, selectedRows) => {
    console.log(
      `selectedRowKeys: ${selectedRowKeys}`,
      "selectedRows: ",
      selectedRows
    );
  },
  onSelect: (record, selected, selectedRows) => {
    record.selected = selected;
    console.log(record, selected, selectedRows);
  },
  onSelectAll: (selected, selectedRows, changeRows) => {
    console.log(selected, selectedRows, changeRows);
  }
};

const style = { display: "block", marginLeft: 0 };

export default class AssignConnectors extends React.Component<
  { completeDataSource: {}[] },
  { searchText: string, dataSource: {}[] }
> {
  state = { searchText: "", dataSource: this.props.completeDataSource };

  columns = [
    {
      title: "Connector",
      dataIndex: "name",
      key: "name",
      sortOrder: "ascend",
      onHeaderCell(column) {
        return { id: column.key };
      },
      sorter(a, b) {
        // TODO try set this to 'ascend' or something
        if (a.name == null) return -1;
        if (b.name == null) return 1;
        if (a.name.toLowerCase() > b.name.toLowerCase()) return -1;
        if (a.name.toLowerCase() < b.name.toLowerCase()) return 1;
        return 0;
      },
      filterMultiple: false,
      filters: [
        { text: "Selected", value: "SELECTED" },
        { text: "Unselected", value: "UNSELECTED" },
        { text: "All", value: "ALL" }
      ],
      onFilter(value, record) {
        if (value === "SELECTED") {
          return record.selected;
        }
        if (value === "UNSELECTED") {
          return !record.selected;
        }
        return true;
      },
      render: text => {
        const matches = match(text, this.state.searchText);
        const splitMatchesResult = splitMatches(text, matches);

        const result = splitMatchesResult.reduce((prev, curr) => {
          return (
            <React.Fragment>
              {prev}
              {curr.isMatch ? <strong>{curr.str}</strong> : curr.str}
            </React.Fragment>
          );
        }, "");
        console.log(splitMatchesResult);
        return result;
        return text;
      }
    }
  ];

  handleSearchTextChange = (e: SyntheticEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget;

    // const dataSource = this.props.completeDataSource.filter(record => {
    //   return record.name.toLowerCase().includes(value.toLowerCase());
    // });

    const dataSource = value
      ? filter(this.props.completeDataSource, value, {
          key: "name"
        })
      : this.props.completeDataSource;

    this.setState({ searchText: value, dataSource });
  };

  render() {
    const { searchText, dataSource } = this.state;

    return (
      <Table
        components={{
          header: {
            cell: props => {
              if (props.id === "name") {
                return (
                  <th
                    style={{
                      display: "flex",
                      alignItems: "center",
                      justifyContent: "space-between"
                    }}
                    {...props}
                  >
                    {props.children}
                    <Input
                      style={{ width: "70%" }}
                      value={searchText}
                      onChange={this.handleSearchTextChange}
                    />
                  </th>
                );
              }
              return <th {...props} />;
            }
          }
        }}
        bordered
        rowSelection={rowSelection}
        columns={this.columns}
        dataSource={dataSource}
        expandedRowRender={record => {
          return (
            <div style={{ display: "flex" }}>
              <Checkbox.Group style={{}}>
                <Checkbox style={style}>
                  Cedula <Icon type="key" />
                </Checkbox>
                <Checkbox style={style}>Nombre</Checkbox>
                <Checkbox style={style}>Apellido</Checkbox>
                <Checkbox style={style}>Direccion</Checkbox>
                <Checkbox style={style}>Estado Civil</Checkbox>
              </Checkbox.Group>

              <div>
                <Divider
                  type="vertical"
                  style={{ height: "80%", top: "10%" }}
                />
              </div>

              <div>
                <strong>Data found</strong>
                <div>
                  <Radio.Group defaultValue={true}>
                    <Radio style={style} value={true}>
                      Deshabilitar
                    </Radio>
                    <Radio style={style} value={false}>
                      Habilitar
                    </Radio>
                  </Radio.Group>
                </div>
              </div>

              <div>
                <strong>No data found</strong>
                <div>
                  <Radio.Group defaultValue={true}>
                    <Radio style={style} value={true}>
                      Allow complete
                    </Radio>
                    <Radio style={style} value={false}>
                      Do not allow complete
                    </Radio>
                  </Radio.Group>
                </div>
              </div>
            </div>
          );
        }}
      />
    );
  }
}
