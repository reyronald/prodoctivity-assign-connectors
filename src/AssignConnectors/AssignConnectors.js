// @flow
import * as React from "react";
import { filter } from "fuzzaldrin-plus";
import update from "immutability-helper";
import { DragDropContext } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import noop from "noop2";
import { highlightChars } from "highlight-matches-utils";
import Table from "antd/lib/table";
import Icon from "antd/lib/icon";
import Input from "antd/lib/input";
import Divider from "antd/lib/divider";
import Checkbox from "antd/lib/checkbox";
import Radio from "antd/lib/radio";
import { css } from "react-emotion";
import DraggableBodyRow from "./DraggableBodyRow";
import { mergeConnectorsAndFormConnectors } from "./utils";
import type { WebServiceConnector, FormConnector } from "../_lib/types";

const anchorTagsStyles = css`
  a {
    cursor: pointer;
    color: #007bff !important;
  }
`;

const checkboxAndRadioStyle = { display: "block", marginLeft: 0 };

const MyDivider = () => (
  <div>
    <Divider type="vertical" style={{ height: "90%", top: "5%" }} />
  </div>
);

export type ConnectorToAssign = {|
  ...FormConnector,
  ...{| key: string, selected?: boolean |}
|};

type P = {
  connectors: WebServiceConnector[],
  formConnectors: FormConnector[],
  onChange: (formConnectors: FormConnector[]) => void,
  onChangeConnectorsToAssign: (connectorsToAssign: ConnectorToAssign[]) => void
};

type S = {
  searchText: string,
  connectors: WebServiceConnector[],
  filteredConnectors: ?(WebServiceConnector[]),
  connectorsToAssign: ConnectorToAssign[]
};

class AssignConnectors extends React.PureComponent<P, S> {
  static displayName = "AssignConnectors";

  static defaultProps = {
    onChangeConnectorsToAssign: noop
  };

  state = {
    searchText: "",
    connectors: this.props.connectors.map(cds => ({
      ...cds,
      key: cds.name
    })),
    filteredConnectors: undefined,
    connectorsToAssign: mergeConnectorsAndFormConnectors(
      this.props.connectors,
      this.props.formConnectors
    )
  };

  notifyChange = () => {
    const formConnectors: FormConnector[] = this.state.connectorsToAssign
      .filter(ac => ac.selected)
      .map(ac => {
        const { key, selected, ...formConnector } = ac;
        return formConnector;
      });
    this.props.onChange(formConnectors);
    this.props.onChangeConnectorsToAssign(this.state.connectorsToAssign);
  };

  constructor(props) {
    super(props);
    this.notifyChange();
  }

  columns = [
    {
      title: "Connector",
      dataIndex: "name",
      key: "name",
      onHeaderCell(column) {
        return { id: column.key };
      },
      filterMultiple: false,
      filters: [
        { text: "Selected", value: "SELECTED" },
        { text: "Unselected", value: "UNSELECTED" },
        { text: "All", value: "ALL" }
      ],
      onFilter: (
        value: "SELECTED" | "UNSELECTED" | "ALL",
        record: WebServiceConnector
      ) => {
        if (value === "ALL") {
          return true;
        }

        const currentRecordIsSelected = this.state.connectorsToAssign.some(
          ({ name, selected }) => name === record.name && selected
        );
        if (value === "SELECTED") {
          return currentRecordIsSelected;
        }
        if (value === "UNSELECTED") {
          return !currentRecordIsSelected;
        }

        throw new Error(`Unrecognized filter value: ${value}`);
      },
      render: text =>
        highlightChars(text, this.state.searchText, (s, i) => (
          <mark
            key={i}
            style={{
              padding: 0,
              fontWeight: 600,
              backgroundColor: "transparent"
            }}
          >
            {s}
          </mark>
        ))
    }
  ];

  getHeaderCell = headerCellProps => {
    if (headerCellProps.id === "name") {
      return (
        <th
          style={{
            display: "flex",
            alignItems: "center"
          }}
          {...headerCellProps}
        >
          {headerCellProps.children}
          <Input
            name="search"
            placeholder="Search..."
            style={{
              flexGrow: 1,
              marginLeft: 40,
              width: "auto"
            }}
            value={this.state.searchText}
            onChange={this.handleSearchTextChange}
          />
        </th>
      );
    }

    return <th {...headerCellProps} />;
  };

  components = {
    body: {
      row: DraggableBodyRow
    },
    header: {
      cell: this.getHeaderCell
    }
  };

  moveRow = (dragIndex: number, hoverIndex: number) => {
    this.setState(prevState => {
      const { connectors, connectorsToAssign } = update(prevState, {
        connectors: {
          $splice: [
            [dragIndex, 1],
            [hoverIndex, 0, prevState.connectors[dragIndex]]
          ]
        },
        connectorsToAssign: {
          $splice: [
            [dragIndex, 1],
            [hoverIndex, 0, prevState.connectorsToAssign[dragIndex]]
          ]
        }
      });

      return { connectors, connectorsToAssign };
    }, this.notifyChange);
  };

  handleSearchTextChange = (e: SyntheticEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget;

    this.setState(prevState => {
      const filteredConnectors = value
        ? filter(prevState.connectors, value, {
            key: "name"
          })
        : undefined;

      return { searchText: value, filteredConnectors };
    });
  };

  rowSelection = {
    onChange: (selectedRowKeys, selectedRows: Array<WebServiceConnector>) => {
      this.setState(prevState => {
        const connectorsToAssign = prevState.connectorsToAssign.map(c => {
          c.selected = selectedRows.some(row => c.name === row.name);
          return c;
        });

        return { connectorsToAssign };
      }, this.notifyChange);
    }
  };

  expandedRowRender = (record: WebServiceConnector) => {
    let index = this.state.connectorsToAssign.findIndex(
      c => c.name === record.name
    );
    const assignedConnector = this.state.connectorsToAssign[index];

    return (
      <div
        style={{
          display: "flex",
          flexWrap: "wrap",
          justifyContent: "space-around"
        }}
      >
        <div>
          {record.inputMapping.map(r => (
            <div key={r.dataElement.name}>
              <Icon type="key" /> {r.dataElement.name}
            </div>
          ))}
        </div>

        <MyDivider />

        <Checkbox.Group
          defaultValue={assignedConnector.output.map(i => i.name)}
          onChange={names => {
            this.setState(prevState => {
              const connectorsToAssign = [...prevState.connectorsToAssign];
              connectorsToAssign[index].output = names.map(name => ({
                name
              }));
              return { connectorsToAssign };
            }, this.notifyChange);
          }}
        >
          {record.outputMapping.filter(r => r.dataElement != null).map(r => (
            <Checkbox
              style={checkboxAndRadioStyle}
              key={r.dataElement.name}
              value={r.dataElement.name}
            >
              {r.dataElement.name}
            </Checkbox>
          ))}
        </Checkbox.Group>

        <MyDivider />

        <div>
          <div style={{ marginBottom: "1rem", marginRight: "2rem" }}>
            <strong>Data found</strong>
            <div>
              <Radio.Group
                name="shouldDisableOnData"
                defaultValue={assignedConnector.shouldDisableOnData}
                onChange={e => {
                  this.setState(prevState => {
                    const connectorsToAssign = [
                      ...prevState.connectorsToAssign
                    ];
                    connectorsToAssign[index].shouldDisableOnData =
                      e.target.value;
                    return { connectorsToAssign };
                  }, this.notifyChange);
                }}
              >
                <Radio style={checkboxAndRadioStyle} value={true}>
                  Disable
                </Radio>
                <Radio style={checkboxAndRadioStyle} value={false}>
                  Enable
                </Radio>
              </Radio.Group>
            </div>
          </div>
          <div>
            <strong>No data found</strong>
            <div>
              <Radio.Group
                name="shouldDisableOnNoData"
                defaultValue={assignedConnector.shouldDisableOnNoData}
                onChange={e => {
                  this.setState(prevState => {
                    const connectorsToAssign = [
                      ...prevState.connectorsToAssign
                    ];
                    connectorsToAssign[index].shouldDisableOnNoData =
                      e.target.value;
                    return { connectorsToAssign };
                  }, this.notifyChange);
                }}
              >
                <Radio style={checkboxAndRadioStyle} value={true}>
                  Disable
                </Radio>
                <Radio style={checkboxAndRadioStyle} value={false}>
                  Enable
                </Radio>
              </Radio.Group>
            </div>
          </div>
        </div>

        <div style={{ flexGrow: 1, display: "flex", flexDirection: "column" }}>
          <div>No data found message:</div>
          <textarea
            defaultValue={assignedConnector.noDataMessage}
            style={{ width: "100%", height: "100%" }}
            onChange={e => {
              const message = e.currentTarget.value;
              this.setState(prevState => {
                const connectorsToAssign = [...prevState.connectorsToAssign];
                connectorsToAssign[index].noDataMessage = message;
                return { connectorsToAssign };
              }, this.notifyChange);
            }}
          />
        </div>
      </div>
    );
  };

  onRow = (record, index) => ({ index, moveRow: this.moveRow });

  render() {
    const { filteredConnectors, connectors, connectorsToAssign } = this.state;

    const selectedRowKeys = connectorsToAssign
      .filter(({ selected }) => selected)
      .map(({ name }) => name);

    return (
      <Table
        className={anchorTagsStyles}
        bordered
        components={this.components}
        rowSelection={{
          selectedRowKeys,
          onChange: this.rowSelection.onChange
        }}
        expandedRowRender={this.expandedRowRender}
        onRow={this.onRow}
        columns={this.columns}
        dataSource={filteredConnectors || connectors}
      />
    );
  }
}

export default DragDropContext(HTML5Backend)(AssignConnectors);
