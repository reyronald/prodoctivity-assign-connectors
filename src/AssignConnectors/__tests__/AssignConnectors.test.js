import React from "react";
import { render } from "react-dom";
import { Simulate } from "react-dom/test-utils";
import noop from "noop2";
import AssignConnectors from "../";

describe("AssignConnectors", () => {
  it("should render properly and call onChange immediately", () => {
    const onChangeConnectorsToAssign = jest.fn();
    const mountNode = document.createElement("div");
    render(
      <AssignConnectors
        connectors={getData()}
        onChange={noop}
        onChangeConnectorsToAssign={onChangeConnectorsToAssign}
      />,
      mountNode
    );

    expect(onChangeConnectorsToAssign).toHaveBeenCalledTimes(1);
    expect(onChangeConnectorsToAssign).toHaveBeenLastCalledWith([
      {
        key: "Social Security Database",
        name: "Social Security Database",
        output: [],
        shouldDisableOnData: true,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      },
      {
        key: "Customers",
        name: "Customers",
        output: [],
        shouldDisableOnData: true,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      },
      {
        key: "People",
        name: "People",
        output: [],
        shouldDisableOnData: true,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      }
    ]);
  });

  it("checking (selecting) connectors should work", () => {
    const onChangeConnectorsToAssign = jest.fn();
    const mountNode = document.createElement("div");
    let assignedConnectors = null;
    render(
      <AssignConnectors
        connectors={getData()}
        onChange={noop}
        onChangeConnectorsToAssign={args => {
          onChangeConnectorsToAssign(args);
          assignedConnectors = args;
        }}
      />,
      mountNode
    );

    const selectionInputs = mountNode.querySelectorAll(
      ".ant-table-selection-column input"
    );
    selectionInputs[1].checked = true;
    Simulate.change(selectionInputs[1]);
    selectionInputs[3].checked = true;
    Simulate.change(selectionInputs[3]);

    const selectedConnectors = assignedConnectors.filter(a => a.selected);
    expect(onChangeConnectorsToAssign).toHaveBeenCalledTimes(3);
    expect(selectedConnectors).toHaveLength(2);
    expect(selectedConnectors.map(sc => sc.name)).toEqual([
      "Social Security Database",
      "People"
    ]);

    selectionInputs[1].checked = false;
    Simulate.change(selectionInputs[1]);
    const selectedConnectors2 = assignedConnectors.filter(a => a.selected);
    expect(onChangeConnectorsToAssign).toHaveBeenCalledTimes(4);
    expect(selectedConnectors2).toHaveLength(1);
    expect(selectedConnectors2.map(sc => sc.name)).toEqual(["People"]);

    /**
     * Testing filters
     */
    const filterIcon = mountNode.querySelector('i[title="Filter menu"]');
    Simulate.click(filterIcon);

    const menuItems = Array.from(
      mountNode.querySelectorAll(".ant-dropdown-menu-item")
    );

    const getMenuItem = value => {
      return menuItems.find(menuItem =>
        Array.from(menuItem.querySelectorAll("span")).some(
          span => span.textContent === value
        )
      );
    };

    // 'Selected' filter
    const selectedMenuItem = getMenuItem("Selected");
    Simulate.click(selectedMenuItem);
    Simulate.click(mountNode.querySelector(".confirm"));
    expect(mountNode.querySelectorAll("tr td:nth-of-type(3)")).toHaveLength(1);

    // 'Unselected' filter
    const unselectedMenuItem = getMenuItem("Unselected");
    Simulate.click(unselectedMenuItem);
    Simulate.click(mountNode.querySelector(".confirm"));
    expect(mountNode.querySelectorAll("tr td:nth-of-type(3)")).toHaveLength(2);

    // 'All' filter
    const allMenuItem = getMenuItem("All");
    Simulate.click(allMenuItem);
    Simulate.click(mountNode.querySelector(".confirm"));
    expect(mountNode.querySelectorAll("tr td:nth-of-type(3)")).toHaveLength(3);

    // TODO get the instance of the component and call onFilter directly
    // to assert what happens when the value is unrecognized (Should throw Error)
  });

  it("fuzzy search should work", () => {
    const mountNode = document.createElement("div");
    render(
      <AssignConnectors connectors={getData()} onChange={noop} />,
      mountNode
    );

    const search = mountNode.querySelector('input[name="search"]');
    search.value = "ssd";
    Simulate.change(search);

    const tableCells = mountNode.querySelectorAll("tr td:nth-of-type(3)");
    const tableCell = tableCells[0];
    const searchedConnector = tableCell.textContent;
    expect(searchedConnector).toBe("Social Security Database");
    expect(tableCells).toHaveLength(1);
    expect(tableCell).toMatchSnapshot();
  });

  it("should call onChange with updated data when UI is interacted with", () => {
    const onChangeConnectorsToAssign = jest.fn();
    const mountNode = document.createElement("div");
    render(
      <AssignConnectors
        connectors={[getData()[0]]}
        onChange={noop}
        onChangeConnectorsToAssign={onChangeConnectorsToAssign}
      />,
      mountNode
    );

    Simulate.click(mountNode.querySelector(".ant-table-row-expand-icon"));

    // Testing `output`
    const checkbox = mountNode.querySelector('input[value="Last Name"]');
    checkbox.checked = true;
    Simulate.change(checkbox);

    expect(onChangeConnectorsToAssign).toHaveBeenCalledTimes(2);
    expect(onChangeConnectorsToAssign).toHaveBeenLastCalledWith([
      {
        key: "Social Security Database",
        name: "Social Security Database",
        output: [{ name: "Last Name" }],
        shouldDisableOnData: true,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      }
    ]);

    // Testing `shouldDisableOnData`
    const shouldDisableOnData = mountNode.querySelector(
      `input[name="shouldDisableOnData"][value="${false}"]`
    );
    shouldDisableOnData.checked = false;
    Simulate.change(shouldDisableOnData);

    expect(onChangeConnectorsToAssign).toHaveBeenLastCalledWith([
      {
        key: "Social Security Database",
        name: "Social Security Database",
        output: [{ name: "Last Name" }],
        shouldDisableOnData: false,
        shouldDisableOnNoData: false,
        noDataMessage: ""
      }
    ]);

    // Testing `shouldDisableOnNoData`
    const shouldDisableOnNoData = mountNode.querySelector(
      `input[name="shouldDisableOnNoData"][value="${true}"]`
    );
    shouldDisableOnNoData.checked = true;
    Simulate.change(shouldDisableOnNoData);
    expect(onChangeConnectorsToAssign).toHaveBeenLastCalledWith([
      {
        key: "Social Security Database",
        name: "Social Security Database",
        output: [{ name: "Last Name" }],
        shouldDisableOnData: false,
        shouldDisableOnNoData: true,
        noDataMessage: ""
      }
    ]);

    // Testing `noDataMessage`
    const textarea = mountNode.querySelector("textarea");
    textarea.value = "No data message";
    Simulate.change(textarea);
    expect(onChangeConnectorsToAssign).toHaveBeenLastCalledWith([
      {
        key: "Social Security Database",
        name: "Social Security Database",
        output: [{ name: "Last Name" }],
        shouldDisableOnData: false,
        shouldDisableOnNoData: true,
        noDataMessage: "No data message"
      }
    ]);
  });

  it("should not crash expanding connector that has no data element in output mapping", () => {
    const onChangeConnectorsToAssign = jest.fn();
    const mountNode = document.createElement("div");
    render(
      <AssignConnectors
        connectors={[
          {
            name: "Social Security Database",
            inputMapping: [{ dataElement: { name: "Identification" } }],
            outputMapping: [
              // Mapping object Doesn't have a "dataElement" prop
              { dataLookupExpression: "name" }
            ]
          }
        ]}
        onChange={noop}
        onChangeConnectorsToAssign={onChangeConnectorsToAssign}
      />,
      mountNode
    );

    Simulate.click(mountNode.querySelector(".ant-table-row-expand-icon"));
  });

  it("dnd", () => {
    const onChange = jest.fn();
    const mountNode = document.createElement("div");
    const data = getData();
    render(
      <AssignConnectors connectors={[data[0], data[1]]} onChange={onChange} />,
      mountNode
    );
  });

  // TODO...
  // https://github.com/atlassian/react-beautiful-dnd/blob/master/test/unit/integration/hooks-integration.spec.js
  // mousedown -> mousemove -> mouseup
  //
});

function getData() {
  return [
    {
      key: 1,
      name: "Social Security Database",
      inputMapping: [{ dataElement: { name: "Identification" } }],
      outputMapping: [
        { dataElement: { name: "First Name" } },
        { dataElement: { name: "Last Name" } },
        { dataElement: { name: "Gender" } }
      ]
    },
    {
      key: 2,
      name: "Customers",
      inputMapping: [{ dataElement: { name: "Customer Name" } }],
      outputMapping: [
        { dataElement: { name: "Identification" } },
        { dataElement: { name: "Address" } },
        { dataElement: { name: "Civil Status" } }
      ]
    },
    {
      key: 3,
      name: "People",
      inputMapping: [{ dataElement: { name: "Person Name" } }],
      outputMapping: [
        { dataElement: { name: "Age" } },
        { dataElement: { name: "Height" } }
      ]
    }
  ];
}
