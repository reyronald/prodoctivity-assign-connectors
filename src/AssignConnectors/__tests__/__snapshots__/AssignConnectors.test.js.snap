// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssignConnectors fuzzy search should work 1`] = `
<td
  class="ant-table-column-has-filters"
>
  <mark
    style="padding: 0px; font-weight: 600; background-color: transparent;"
  >
    S
  </mark>
  ocial 
  <mark
    style="padding: 0px; font-weight: 600; background-color: transparent;"
  >
    S
  </mark>
  ecurity 
  <mark
    style="padding: 0px; font-weight: 600; background-color: transparent;"
  >
    D
  </mark>
  atabase
</td>
`;
